core/libp2p/host_builder.cpp

```diff
@@ -2,11 +2,173 @@
  * Copyright Soramitsu Co., Ltd. All Rights Reserved.
  * SPDX-License-Identifier: Apache-2.0
  */
- 
+
 #include "libp2p/host_builder.hpp"
 
+#include <exception>
+
+#include "libp2p/crypto/key_generator/key_generator_impl.hpp"
+#include "libp2p/crypto/random_generator/boost_generator.hpp"
+#include "libp2p/crypto/random_generator/std_generator.hpp"
+#include "libp2p/dht/dht_impl.hpp"
+#include "libp2p/discovery/discovery_impl.hpp"
+#include "libp2p/muxer/yamux.hpp"
+#include "libp2p/peer/address_repository/inmem_address_repository.hpp"
+#include "libp2p/peer/key_repository/inmem_key_repository.hpp"
+#include "libp2p/peer/protocol_repository/inmem_protocol_repository.hpp"
+#include "libp2p/routing/routing_impl.hpp"
+#include "libp2p/security/security_impl.hpp"
+#include "libp2p/transport/impl/transport_impl.hpp"
+
+namespace {
+  /**
+   * Check if both keys in the keypair have data and type set
+   * @param keypair to be checked
+   * @return true, if keypair is well-formed, false otherwise
+   */
+  bool keypairIsWellFormed(const libp2p::crypto::KeyPair &keypair) {
+    using KeyType = libp2p::crypto::Key::Type;
+
+    const auto &pubkey = keypair.publicKey;
+    const auto &privkey = keypair.privateKey;
+    return !pubkey.data.empty() && pubkey.type != KeyType::UNSPECIFIED
+        && !privkey.data.empty() && privkey.type != KeyType::UNSPECIFIED;
+  }
+}  // namespace
+
 namespace libp2p {
+  using multi::Multiaddress;
+
+  HostBuilder::HostBuilder(boost::asio::io_context &io_context)
```

Comment:
do not use it here, as boost::asio::io_context is impl detail. 

users will be creating it as 
```
auto tcp = new Tcp(context);
builder.addTransport(tcp);
```

Reply:
How to create a default transport, if the user did not specify it, then?

Reply:
hmm.

Reply:
My other thought was to place `io_context` into `config_` as well so that it's default created and then can be used during the construction of `Transport`. But, this approach makes sense if `Host` will also use the `context`, and I am not sure it will